// appcontroller/appController.js
/**
 * appController
 *
 *
 *
 *
 * @date 21.11.14 - 00:35
 *
 */


var AppClient = function ( socket,id) {
    this.socket = socket;
    this.id = id;
}


appController = {
    preferences:null,
    io:null,
    app:null,
    express:null,
    http:null,
    dataController:null,
    messageType:{
        connection:"c",
        normal:"n"
    },
    clients: [],
    init:function(preferences,dataController,io,app,express){
        this.preferences = preferences;
        this.io = io;
        this.app = app;
        this.express = express;

        //Data Controller
        this.dataController = dataController;

    },
    messageRecieved:function(socket,msg){
        //Connection
        if(msg.t==appController.messageType.connection){
            appController.clientConnected(socket,msg.clientid);

            socket.emit('message', {t: "cb",callback:msg.data.callback});

            //Normal Message
        }else if(msg.t==appController.messageType.normal){
            var client = appController.getClient(socket,msg.clientid);
            if(client&&client.length>0)
                this.dataController.handleMessage(this,client[0],msg.data);
        }
    }
    ,clientConnected: function (socket, clientid) {
        console.log("App: Connected")

        var client = new AppClient(socket,clientid);
        this.removeClient(client);
        this.handleFutureDisconnect(client);
        appController.clients.push(client);



    },
    getClient:function(socket,id){
        return appController.clients.filter(function (element) {
            return (!socket||element.socket == socket)&&(!id||element.id == id)
        });
    }
    ,
    removeClient: function (client) {
        //Remove Element from List
        appController.clients = appController.clients.filter(function (element) {
            return (element.socket != client.socket)
        });

    },
    handleFutureDisconnect: function (client) {
        client.socket.on('disconnect', function () {
            appController.clientDisonnected(client);
        });
    },

    clientDisonnected: function (client) {
        console.log("App: Disconnected")
        this.removeClient(client.socket);

    }


}


module.exports = appController;
